snakemake/executors/flux.py

```diff
@@ -0,0 +1,266 @@
+            "{exception.note} {exception.type} {result} {runtime} {status}"
+            "{ranks:h} {t_remaining} {annotations}"
+        )
+        self.jobs_formatter = flux.job.JobInfoFormat(jobs_format)
+
+        # Note there is no attr for "id", its always returned
+        fields2attrs = {
+            "id.f58": (),
+            "username": ("userid",),
+            "exception.occurred": ("exception_occurred",),
+            "exception.type": ("exception_type",),
+            "exception.note": ("exception_note",),
+            "runtime": ("t_run", "t_cleanup"),
+            "status": ("state", "result"),
+            "status_abbrev": ("state", "result"),
+            "t_remaining": ("expiration", "state", "result"),
+        }
+
+        # Set job attributes we will use later to get job statuses
+        self.job_attrs = set()
+        for field in self.jobs_formatter.fields:
+            if field not in fields2attrs:
+                self.job_attrs.update((field,))
+            else:
+                self.job_attrs.update(fields2attrs[field])
+
+    def _init_flux(self):
+        """
+        Ensure we can import flux in the environment.
+        """
+        if not flux:
+            sys.exit(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
+            )
+        self._fexecutor = flux.job.FluxExecutor()
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        mem_mb = job.resources.get("mem_mb", 15360)
+
+        # Update default resources
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+        self.default_resources.set_resource("mem_mb", mem_mb)
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile.replace(self.workdir, "").strip(os.sep)
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def _generate_environment(self):
+        """
+        Loop through envvars (keys to host environment) and add
+        any that are requested for the runtime environment
+        """
+        envvars = dict(os.environ)
+        for key in self.envvars:
+            try:
+                envvars[key] = os.environ[key]
+            except Exception:
+                continue
```

Comment:
This was recommended by a flux developer - I suspect if there is a non zero exit code it raises an exception, possibly this one: https://github.com/flux-framework/flux-core/blob/0fa713c19b1f51b76d9f74063657b9fd54f8e60f/src/bindings/python/flux/job/executor.py#L248-L257.

We could try changing to RuntimeError to be more specific - I'm not practiced enough with flux to know if it can throw other kinds of exceptions.

Reply:
okay changed to RuntimeError to reflect what I see in the flux code.

Reply:
Mhm, I still don't see how there could possibly a runtime error when accessing os.environ.

Reply:
Sorry - the RuntimeError is here https://github.com/snakemake/snakemake/pull/1810/files#diff-b201d1534feeee9b245773a64a6e98ab170f984d4a7ddd0def068448ceaf3273R232 when we get a future result (I associated the comment with the wrong spot). For the above, I originally had the environment empty (and think maybe it should still be that way?) as opposed to starting with the entire os.environ - I forgot to ask about that. For a cluster-like executor should we always be including the user environment, or only a subset of variables requested (e.g., if we did:

```
        envvars = {}
        for key in self.envvars:
            try:
                envvars[key] = os.environ[key]
            except Exception:
                continue
```
that's where the exception could be needed - because perhaps something requested in self.envvars isn't acutally in the environment. Buit maybe I'm misunderstanding the expected content of self.envvars? Sorry it's been a while!

Reply:
Snakemake ensures that everything in self.envvars is actually present in os.environ before the executor is invoked. Hence, you can safely assume that no error happens. And if so, we actually want to see that exception because that would be a bug in Snakemake. Hence, all you need should be
```
envvars = {os.environ[key] for key in self.envvars}
```

Reply:
I've changed this to provide flux the entire user env - there are errors finding snakemake, etc. without that.